lib/SILOptimizer/Utils/Generics.cpp

```diff
@@ -44,36 +121,54 @@ static unsigned getBoundGenericDepth(Type t) {
   return Depth;
 }
 
+/// For a dependent type of a generic type returns its base
+/// generic type. For a generic type parameter type, returns this
+/// type.
+GenericTypeParamType *getBaseGenericTypeParamType(CanType Ty) {
```

Comment:
Make this take a Type instead of CanType and canonicalize the result instead of each iteration